File: Completed/Go/199.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
import (
	"container/list"
)

type node struct {
	level int
	ele   *TreeNode
}

func rightSideView(root *TreeNode) []int {
	q, maxDepth := list.New(), 0
	if root != nil {
		q.PushBack(node{0, root})
	}

	m := map[int]int{}

	for q.Len() > 0 {
		n := q.Front().Value.(node)
		maxDepth = max(maxDepth, n.level)

		m[n.level] = n.ele.Val

		if n.ele.Left != nil {
			q.PushBack(node{n.level + 1, n.ele.Left})
		}
		if n.ele.Right != nil {
			q.PushBack(node{n.level + 1, n.ele.Right})
		}
		q.Remove(q.Front())
	}

	if len(m) == 0 {
		return []int{}
	}

	res := make([]int, maxDepth+1)
	for lvl, val := range m {
		res[lvl] = val
	}

	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: Completed/python3/199.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
      result = []
      self.helper(root, result)
      return result
    def helper(self, root: TreeNode, lst: List[int], level: int = 0):
      if root == None:
        return
      
      if level >= len(lst):
        lst.append(root.val)
      else:
        lst[level]=root.val
      
      self.helper(root.left, lst, level+1)
      self.helper(root.right, lst, level+1)
        
